js/compiler/compiler.go

```diff
@@ -44,103 +44,114 @@ var (
+type CompatibilityMode uint8
 
-	// JS pointers.
-	this      goja.Value
-	transform goja.Callable
-	mutex     sync.Mutex //TODO: cache goja.CompileAST() in an init() function?
-}
+const (
+	// CompatibilityModeExtended achieves ES6+ compatibility with Babel and core.js
+	CompatibilityModeExtended CompatibilityMode = iota + 1
+	// CompatibilityModeBase is standard goja ES5.1+
+	CompatibilityModeBase
+)
 
-// Constructs a new compiler.
-func New() (*Compiler, error) {
-	var err error
-	once.Do(func() {
-		compilerInstance, err = new()
-	})
+// A Compiler compiles JavaScript source code (ES5.1 or ES6) into a goja.Program
+type Compiler struct{}
 
-	return compilerInstance, err
+// New returns a new Compiler
+func New() *Compiler {
+	return &Compiler{}
 }
 
-func new() (*Compiler, error) {
-	conf := rice.Config{
-		LocateOrder: []rice.LocateMethod{rice.LocateEmbedded},
+// Transform the given code into ES5, while synchronizing to ensure only a single
+// babel instance is in use.
+func (c *Compiler) Transform(src, filename string) (code string, srcmap SourceMap, err error) {
+	var b *babel
+	if b, err = newBabel(); err != nil {
+		return
 	}
 
-	babelSrc := conf.MustFindBox("lib").MustString("babel.min.js")
+	b.mutex.Lock()
+	defer b.mutex.Unlock()
+	return b.Transform(src, filename)
+}
 
-	c := &Compiler{vm: goja.New()}
-	if _, err := c.vm.RunString(babelSrc); err != nil {
-		return nil, err
+// Compile the program in the given CompatibilityMode, optionally running pre and post code.
+func (c *Compiler) Compile(src, filename string, pre, post string,
+	strict bool, compatMode CompatibilityMode) (*goja.Program, string, error) {
+	code := pre + src + post
+	ast, err := parser.ParseFile(nil, filename, code, 0)
+	if err != nil {
+		if compatMode == CompatibilityModeExtended {
+			code, _, err = c.Transform(src, filename)
+			if err != nil {
+				return nil, code, err
+			}
+			return c.Compile(code, filename, pre, post, strict, compatMode)
+		}
+		return nil, src, err
 	}
+	pgm, err := goja.CompileAST(ast, strict)
+	return pgm, code, err
+}
 
-	c.this = c.vm.Get("Babel")
-	thisObj := c.this.ToObject(c.vm)
-	if err := c.vm.ExportTo(thisObj.Get("transform"), &c.transform); err != nil {
-		return nil, err
-	}
+type babel struct {
+	vm        *goja.Runtime
+	this      goja.Value
+	transform goja.Callable
+	mutex     sync.Mutex //TODO: cache goja.CompileAST() in an init() function?
+}
+
+func newBabel() (*babel, error) {
+	var err error
 
-	return c, nil
+	once.Do(func() {
+		conf := rice.Config{
+			LocateOrder: []rice.LocateMethod{rice.LocateEmbedded},
+		}
+		babelSrc := conf.MustFindBox("lib").MustString("babel.min.js")
+		vm := goja.New()
+		if _, err = vm.RunString(babelSrc); err != nil {
+			return
+		}
+
+		this := vm.Get("Babel")
+		bObj := this.ToObject(vm)
+		babl = &babel{vm: vm, this: this}
+		if err = vm.ExportTo(bObj.Get("transform"), &babl.transform); err != nil {
+			return
+		}
+	})
```

Comment:
If it wasn't for the tests I would argue we can drop the whole global for  babel ... 
Can you drop it and run all the tests (with -count 1) with and without and tell how big of difference does it make ?

Reply:
> how big of difference does it make ?

As far as run time goes, quite substantial for `./js`:
- `ok      github.com/loadimpact/k6/js     51.831s` (with global, shared Babel VM)
- `ok      github.com/loadimpact/k6/js     190.796s` (without global, each test allocates a new Babel VM)

We could consider other alternatives that would get rid of the global and minimize this overhead (e.g. passing a pre-initialized Goja VM to the compiler), but those would fundamentally change how the compiler works and would be out of scope for this PR. The changes done here simply shift the object used as singleton and encapsulate Babel a bit better.

Reply:
Yeah I agree that this is outside the scope of the PR and if the increase is 4x we are not just changing it ...